src/OrchardCore.Modules/OrchardCore.Users/Controllers/EmailAuthenticatorController.cs

```diff
@@ -0,0 +1,217 @@
+        _smtpService = smtpService;
+        _liquidTemplateManager = liquidTemplateManager;
+        _htmlEncoder = htmlEncoder;
+    }
+
+    [Admin, Authorize]
+    public async Task<IActionResult> Index()
+    {
+        var user = await UserManager.GetUserAsync(User);
+        if (user == null)
+        {
+            return UserNotFound();
+        }
+
+        if (!await UserManager.IsEmailConfirmedAsync(user))
+        {
+            return RedirectToAction(nameof(Index));
+        }
+
+        return View();
+    }
+
+    [HttpPost, ValidateAntiForgeryToken, Admin, Authorize]
+    public async Task<IActionResult> RequestCode()
+    {
+        var user = await UserManager.GetUserAsync(User);
+        if (user == null)
+        {
+            return UserNotFound();
+        }
+
+        if (!await UserManager.IsEmailConfirmedAsync(user))
+        {
+            return RedirectToAction(nameof(Index));
+        }
+
+        var code = await UserManager.GenerateEmailConfirmationTokenAsync(user);
+
+        var setings = (await SiteService.GetSiteSettingsAsync()).As<EmailAuthenticatorLoginSettings>();
+        var message = new MailMessage()
+        {
+            To = await UserManager.GetEmailAsync(user),
+            Subject = await GetSubjectAsync(setings, user, code),
+            Body = await GetMessageAsync(setings, user, code),
+            IsHtmlBody = true,
+        };
+
+        var result = await _smtpService.SendAsync(message);
+
+        if (result.Succeeded)
+        {
+            await Notifier.SuccessAsync(H["We have successfully sent an verification code to your email. Please retrieve the code from your email and enter it below."]);
+        }
+        else
+        {
+            await Notifier.ErrorAsync(H["We are unable to send you an email at this time. Please try again later."]);
+
+            return RedirectToAction(nameof(Index));
+        }
+
+        var model = new EnableEmailAuthenticatorViewModel();
+
+        return View(model);
+    }
+
+    [HttpPost, ValidateAntiForgeryToken, Authorize, Admin]
+    public async Task<IActionResult> ValidateCode(EnableEmailAuthenticatorViewModel model)
+    {
+        var user = await UserManager.GetUserAsync(User);
+        if (user == null)
+        {
+            return UserNotFound();
+        }
+
+        if (!ModelState.IsValid)
+        {
+            return View(model);
+        }
+
+        var result = await UserManager.ConfirmEmailAsync(user, StripToken(model.Code));
+
+        if (result.Succeeded)
+        {
+            await EnableTwoFactorAuthenticationAsync(user);
+
+            await Notifier.SuccessAsync(H["Your email has been confirmed."]);
+
+            return await RedirectToTwoFactorAsync(user);
+        }
+
+        await Notifier.ErrorAsync(H["Unable to confirm your email. Please try again."]);
+
+        return View(nameof(RequestCode), model);
+    }
+
+    [IgnoreAntiforgeryToken, Produces("application/json"), AllowAnonymous]
```

Comment:
I think this should be a POST and valide the antiforgery token

Reply:
changed to post